source/UberCore/Authentication/LoginManager.swift

```diff
@@ -261,34 +273,69 @@ import SafariServices
+
+        let authenticators = authenticationProvider.authenticators(for: loginType)
+        executeNativeAuthenticators(authenticators: authenticators) { (fallback: Bool) in
+            if (!fallback) {
                 return
             }
-            
+
             // If we can't open the deeplink, fallback.
             // Privileged scopes require auth code flow.
             // Since that requires server support, fallback to app store if not available.
-            if authenticator.scopes.contains(where: { $0.scopeType == .privileged }) {
+            if authenticationProvider.scopes.contains(where: { $0.scopeType == .privileged }) {
                 if (Configuration.shared.useFallback) {
                     self.loginType = .authorizationCode
                 } else {
-                    let appstoreDeeplink = AppStoreDeeplink(userAgent: nil)
-                    appstoreDeeplink.execute(completion: { _ in
-                        self.loginCompletion(accessToken: nil, error: UberAuthenticationErrorFactory.errorForType(ridesAuthenticationErrorType: .unableToPresentLogin))
-                    })
+                    if let uberProductType = authenticationProvider.productFlowPriority.first?.uberProductType {
+                        let appStoreDeeplink: BaseDeeplink
+                        switch uberProductType {
+                        case .rides:
+                            appStoreDeeplink = RidesAppStoreDeeplink(userAgent: nil)
+                        case .eats:
+                            appStoreDeeplink = EatsAppStoreDeeplink(userAgent: nil)
+                        }
+
+                        appStoreDeeplink.execute(completion: { _ in
+                            self.loginCompletion(accessToken: nil, error: UberAuthenticationErrorFactory.errorForType(ridesAuthenticationErrorType: .unableToPresentLogin))
+                        })
+                    }
                     return
                 }
             } else { // Otherwise fallback to implicit flow
                 self.loginType = .implicit
             }
-            self.login(requestedScopes: authenticator.scopes, presentingViewController: presentingViewController, completion: self.postCompletionHandler)
-        })
+
+
+            self.login(requestedScopes: authenticationProvider.scopes, presentingViewController: presentingViewController, completion: self.postCompletionHandler)
+        }
+    }
+
+    private func executeNativeAuthenticators(authenticators: [UberAuthenticating], completion: ((_ fallback: Bool) -> Void)) {
+        var stop: Bool = false
+        var fallback: Bool = false
+        
+        for authenticator in authenticators where stop == false {
+            self.authenticator = authenticator
+            fallback = false
+
+            DeeplinkManager.shared.open(authenticator.authorizationURL, completion: { error in
+                if error == nil {
+                    return stop = true
```

Comment:
I'm surprised this is valid swift. This callback should have a returntype of void. Also, your code seems to assume that this runs synchronously. I don't think that's a guarantee that is valid.